ccm_web/client/src/models/canvas.ts

```diff
@@ -54,47 +56,128 @@ export class CanvasCourseSectionSort_UserCount implements ICanvasCourseSectionSo
   }
 }
 
-export enum CanvasRoleType {
+export enum CanvasEnrollmentType {
   Student = 'StudentEnrollment',
   Teacher = 'TeacherEnrollment',
   TA = 'TaEnrollment',
   Observer = 'ObserverEnrollment',
   Designer = 'DesignerEnrollment'
 }
 
-export enum ClientRoleType {
+export enum ClientEnrollmentType {
   Student = 'student',
   Teacher = 'teacher',
   TA = 'ta',
   Observer = 'observer',
   Designer = 'designer'
 }
-const clientStringValues = Object.values(ClientRoleType).map(m => String(m))
+const clientStringValues = Object.values(ClientEnrollmentType).map(m => String(m))
 
 export interface CanvasEnrollment {
   id: number
   course_id: number
   course_section_id: number
   user_id: number
-  type: CanvasRoleType
+  type: CanvasEnrollmentType
+}
+
+const clientToCanvasRoleMap: Record<ClientEnrollmentType, CanvasEnrollmentType> = {
+  student: CanvasEnrollmentType.Student,
+  teacher: CanvasEnrollmentType.Teacher,
+  ta: CanvasEnrollmentType.TA,
+  observer: CanvasEnrollmentType.Observer,
+  designer: CanvasEnrollmentType.Designer
 }
 
-const clientToCanvasRoleMap: Record<ClientRoleType, CanvasRoleType> = {
-  student: CanvasRoleType.Student,
-  teacher: CanvasRoleType.Teacher,
-  ta: CanvasRoleType.TA,
-  observer: CanvasRoleType.Observer,
-  designer: CanvasRoleType.Designer
+interface AddableRoleData {
+  addable: true
+  rank: number
+  canvasName: CanvasEnrollmentType
+  clientName: ClientEnrollmentType
 }
 
-export const isValidRole = (role: string): role is ClientRoleType => {
+interface NotAddableRoleData {
+  rank: number
+  addable: false
+}
+
+type RoleData = AddableRoleData | NotAddableRoleData
+
+type CanvasRoleData = Record<RoleEnum, RoleData>
+
+export const AllCanvasRoleData: CanvasRoleData = {
+  StudentEnrollment: {
+    rank: 0,
+    addable: true,
+    canvasName: CanvasEnrollmentType.Student,
+    clientName: ClientEnrollmentType.Student
+  },
+  Observer: {
+    rank: 1,
+    addable: true,
+    canvasName: CanvasEnrollmentType.Observer,
+    clientName: ClientEnrollmentType.Observer
+  },
+  TaEnrollment: {
+    rank: 2,
+    addable: true,
+    canvasName: CanvasEnrollmentType.Observer,
+    clientName: ClientEnrollmentType.TA
+  },
+  DesignerEnrollment: {
+    rank: 3,
+    addable: true,
+    canvasName: CanvasEnrollmentType.Designer,
+    clientName: ClientEnrollmentType.Designer
+  },
+  TeacherEnrollment: {
+    rank: 4,
+    addable: true,
+    canvasName: CanvasEnrollmentType.Teacher,
+    clientName: ClientEnrollmentType.Teacher
+  },
+  'Sub-Account Admin': {
+    rank: 5,
+    addable: false
+  },
+  'Account Admin': {
+    rank: 6,
+    addable: false
+  },
+  'Tool Installer (by ITS Approval only)': {
+    rank: -1,
+    addable: false
+  },
+  'Support Consultant': {
+    rank: -1,
+    addable: false
+  },
+  Librarian: {
+    rank: -1,
+    addable: false
+  },
+  Assistant: {
+    rank: -1,
+    addable: false
+  },
+  Grader: {
+    rank: -1,
+    addable: false
+  }
+} as const
```

Comment:
I like the way you designed this role assigning object. 
<img width="435" alt="Screen Shot 2021-12-17 at 4 51 53 PM" src="https://user-images.githubusercontent.com/8579775/146612438-f9fa8743-0624-4e2a-bef2-e8498b36fa00.png">


Few violation when populating roles list based on canvas role
1. Teacher/Designer should able to add person with teacher role
2.  Assistant has no roles populated in dropdown

Reply:
Thanks, based on your feedback though, I may need to adjust the logic. Because teachers are also allowed to add teachers, the `rank` system doesn't quite work. I may just change it to having a few different role sets and assigning them manually to each of the roles that can add.

Reply:
Okay, this is what I arrived at, but I'll review it again to make sure it makes sense: https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/d94e518dc364d3e0b646a0110a31c7466e1f841e